executor/admin.go

```diff
@@ -344,18 +365,39 @@ func (e *RecoverIndexExec) fetchRecoverRows(ctx context.Context, srcResult dists
 			if result.scanRowCount >= int64(e.batchSize) {
 				return e.recoverRows, nil
 			}
-			handle := kv.IntHandle(row.GetInt64(handleIdx))
-			idxVals := extractIdxVals(row, e.idxValsBufs[result.scanRowCount], e.colFieldTypes)
+			handle, err := e.buildHandleFromChunkRow(row, handleColIndex)
+			if err != nil {
+				return nil, err
+			}
+			idxVals := extractIdxVals(row, e.idxValsBufs[result.scanRowCount], e.colFieldTypes, handleColIndex)
 			e.idxValsBufs[result.scanRowCount] = idxVals
 			e.recoverRows = append(e.recoverRows, recoverRows{handle: handle, idxVals: idxVals, skip: false})
 			result.scanRowCount++
-			result.nextHandle = handle.Next()
+			result.currentHandle = handle.Next()
 		}
 	}
 
 	return e.recoverRows, nil
 }
 
+func (e *RecoverIndexExec) buildHandleFromChunkRow(row chunk.Row, handleColIndex int) (kv.Handle, error) {
+	tblInfo := e.table.Meta()
+	if !tblInfo.IsCommonHandle {
+		return kv.IntHandle(row.GetInt64(handleColIndex)), nil
+	}
+	colInfos := e.columns
+	handleData := make([]types.Datum, 0, len(colInfos)-handleColIndex)
+	for i := handleColIndex; i < len(colInfos); i++ {
+		handleData = append(handleData, row.GetDatum(i, &colInfos[i].FieldType))
+	}
+	var handleBytes []byte
+	handleBytes, err := codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, nil, handleData...)
```

Comment:
The common handle datums may need to truncate, so I build `HandleCols` interface type and use that to build handle.